Improve message loop

HashidaTKS opened the pull request:
# Which issue(s) this PR fixes:

https://github.com/ThinBridge/Chronos/issues/155

# What this PR does / why we need it:

Why we need this:

The redirection is slow in the some pages[1]. 
That is caused by a lack of the implementation of `OnScheduleMessagePumpWork()`.

^1: I can't provide the site to reproduce because it is the internal site for a company.

What this PR does:

* Implement `OnScheduleMessagePumpWork()`
* Add the `MessageLoopWorker` class to manage the timer for `CefDoMessageLoopWork()`
  * `MessageLoopWorker` works on a special message window.
* Remove `Sazabi::PumpMessage()` because it is no longer needed.
* Move process to read leftover messages to `MessageLoopWorker`  from `CSazabi`

`MessageLoopWorker` does:

* Manage the timer so that `DoMessageLoopWork()` executed at at least 30fps.
* For `WM_SCHEDULE_CEF_WORK` message:
  * If `delayMs` is less than or equal to 0, execute `DoMessageLoopWork()` immediately. The timer does not reset.
  * If `delayMs` is greater than 0, set the timer for the specified time. Existing timers will be reset.
* For time timeout message:
  * Reset the timer and execute `DoMessageLoopWork()`.

# How to verify the fixed issue:

## The steps to verify:

1. Open the site with slow redirection.
2. Check if other sites work fine.
    * Full testing will be performed as an integration test. Check some cases at this PR.
3. Check if the Chronos components like the menu bar or the setting dialog and so on work fine.
    * Full testing will be performed as an integration test. Check some cases at this PR.
4. Check if shortcut keys like Ctrl + N wrok fine.
    * Full testing will be performed as an integration test. Check some cases at this PR.

## Expected result:

> 1. Open the site with slow redirection.

The redirection is as fast as v13.1.112.0.

> 2. Check if other sites work fine.

They work fine.

> 3. Check if the Chronos components like the menu bar or the setting dialog and so on work fine.

They work fine.

> 4. Check if shortcut keys like Ctrl + N work fine.

They work fine.

HashidaTKS committed:
message loop: implement OnScheduleMessagePumpWork

Currently, delay_ms is ignored , and CefDoMessageLoopWork() is executed immediately.

ashie reviewed (changes requested):
ついでに、このメッセージが来たときだけ`CefDoMessageLoopWork()`が呼ばれるように`CSazabi::PumpMessage()`の方も改変してみてもらえませんか？たぶん動作に問題はないと思います。 #155 の調査の際、`OnScheduleMessagePumpWork()`が呼ばれてないのに`CefDoMessageLoopWork()`を無駄に呼んでいるケースもそれなりにありました。負荷に大した影響は無いでしょうが、呼ばなくて済むならそれに越したことは無いと思います。

あとちゃんとやるなら`delay_ms`が正の値のときにタイマーを仕込むべきなのですが、 https://github.com/ThinBridge/Chronos/issues/155#issuecomment-1990992267 に書いたように実質的にはあまり意味が無さそうなので、このPRでは必ずしも無理に実装しなくても良いかな、という気分でいます。

ashie commented:
> あとちゃんとやるなら`delay_ms`が正の値のときにタイマーを仕込むべきなのですが、 [#155 (comment)](https://github.com/ThinBridge/Chronos/issues/155#issuecomment-1990992267) に書いたように実質的にはあまり意味が無さそうなので、このPRでは必ずしも無理に実装しなくても良いかな、という気分でいます。

あー、でもこれ私もCEFのサンプルもなんか誤解してるかもしれない。ちゃんとタイマーを仕込んでないと、`CefDoMessageLoopWork()`を呼ぶべきタイミングなのに呼ばれないケースがあるかもしれません。

ashie commented:
> あー、でもこれ私もCEFのサンプルもなんか誤解してるかもしれない。ちゃんとタイマーを仕込んでないと、`CefDoMessageLoopWork()`を呼ぶべきタイミングなのに呼ばれないケースがあるかもしれません。

これ、Chromium側の実装としては以下の`MessagePumpForUI::ScheduleDelayedWork()`で

https://github.com/chromium/chromium/blob/cea674fbaa64e7050e90fc04d1c806f198b34ce3/base/message_loop/message_pump_win.cc#L128-L151

CEF側でこれを継承して`OnScheduleMessagePumpWork()`を呼び出してます。なのでChromium側でタイマーを仕込んでいるのですが、その意味するところをCEFのサンプルも誤ってるかも（CEFのサンプルでは1/30秒以上のタイマーは1/30秒に強制的に縮めてるけど）。

ashie commented:
私が観測したときには、500ms前後の値が飛んでくることがありました。
（条件はよくわからないけど https://www.google.com/ からどこかのリンクに飛んだときとか）

ashie commented:
> CEF側でこれを継承して`OnScheduleMessagePumpWork()`を呼び出してます。なのでChromium側でタイマーを仕込んでいるのですが、その意味するところをCEFのサンプルも誤ってるかも（CEFのサンプルでは1/30秒以上のタイマーは1/30秒に強制的に縮めてるけど）。

例えばJavaScriptで500msでタイマーを仕込んでいて、コンテンツがそのタイミングで再描画したいのに、今のChronosみたいにたまたまそのタイミングでウインドウメッセージが飛んでこない限り再描画できない（今回のコンテンツのように）というのは多分まずいと思います。確実に発火するように`SetTimer`とかで再度メッセージを飛ばさないとだめだと思います。

ashie commented:
> ついでに、このメッセージが来たときだけ`CefDoMessageLoopWork()`が呼ばれるように`CSazabi::PumpMessage()`の方も改変してみてもらえませんか？

もう一点付け加えると、これをやるとたぶん`CSazabi::PumpMessage()`の中の現状のワークアラウンドがほとんどいらなくなって、逆に実装がすっきりするんじゃないかと思います。

HashidaTKS commented:
> 例えばJavaScriptで500msでタイマーを仕込んでいて、コンテンツがそのタイミングで再描画したいのに、今のChronosみたいにたまたまそのタイミングでウインドウメッセージが飛んでこない限り再描画できない（今回のコンテンツのように）というのは多分まずいと思います。確実に発火するようにSetTimerとかで再度メッセージを飛ばさないとだめだと思います。

いずれにしろ待機時間分経過後にどこかで`OnScheduleMessagePumpWork`（`CefDoMessageLoopWork`）が呼ばれるだろうと思ったのですが、確かに何か明示的にユーザーが操作しないと中々次の`OnScheduleMessagePumpWork`が呼ばれなくて、今回と同様の事象が発生するケースがありそうな気がしてきました。

HashidaTKS commented:
タイマーを実装する前に、メッセージが`WM_DO_CEF_WORK`の時だけ`CefDoMessageLoopWork`を実行するように修正したのですが、画面がフリーズして正しく動かなくなりました。
メッセージの送り先がメインフレームで、メインフレーム作成前に`CefInitialize`しているのが問題かと思い、メインフレーム作成後に`CefInitialize`するようにしたのですが、やはりだめでした。
最初のうちは`WM_DO_CEF_WORK`が飛んでくる（`OnScheduleMessagePumpWork`が実行されている）のですが、途中から飛んでこなくなります。

HashidaTKS commented:
というわけでして、余計な`CefDoMessageLoopWork`が実行されないようにするのは別のPRで検証して、このPRではタイマーの実装までとしたいのですがいかがでしょうか？
（タイマー周りは現在実装中です。）

ashie commented:
ではそれでお願いします
CEFのサンプルでも終了時に`CefDoMessageLoopWork()`を少し余計に回すとかの処理があったと思うので、そんなに理想的には行かないのかもしれません。

HashidaTKS commented:
複数のdelay要求がきた時にどのような動作をすべきかよくわかっていなかったが、

[[[https://cefsharp.github.io/api/86.0.x/html/M_CefSharp_IBrowserProcessHandler_OnScheduleMessagePumpWork.htm](https://cef-builds.spotifycdn.com/docs/119.1/classCefBrowserProcessHandler.html#ad17f44d446325a34a3248647df15f842)

> If |delay_ms| is > 0 then the call should be scheduled to happen after the specified delay and any currently pending scheduled call should be cancelled.

といっているので、タイマーについては一番最後にやってきた0より大きいdelay_msが優先される。
CEFのサンプルだと、IsTimerPending状態だとスキップされるので、どうするのが理想的なのかよくわからなかった。
（というより、保留中というのが何をさしているのかわからなかった。）

HashidaTKS committed:
message loop: implement SazabiCefMassageLoopWorker

SazabiCefMassageLoopWorker provides an unified CEF massage loop method
and managements a timer for scheduled message loop works.

HashidaTKS commented:
[main_message_loop_external_pump_win.cc](https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump_win.cc)
[main_message_loop_external_pump.cc](https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc)

を参考にタイマーを実装。単にタイマーを実装するのではなく、`SazabiCefMassageLoopWorker`というクラスを作成し、そこでCEFのメッセージループの処理の統一をしている。

最後に設定された遅延時間後に`DoMessageLoopWork`が実行されるよう、`Sazabi::PumpMessage()`で以下の処理を行う。

1. 受け取ったメッセージが`WM_SCHEDULE_CEF_WORK`だった場合:
    * `delayMs`が0以下なら直ちに`DoMessageLoopWork`を実行。タイマーはリセットしない。
    * `delayMs`が0より大きいなら、指定の時間でタイマーを設定。既存のタイマーはリセット。
2. 受け取ったメッセージがタイマーのタイムアウトだった場合:
    * タイマーをリセットし、`DoMessageLoopWork()`を実行。
3. それ以外のメッセージの場合、`DoMessageLoopWork()`を実行。タイマーはリセットしない。

細かな気になっている点などは別途ソースコード差分へのコメントとして記載。

HashidaTKS committed:
message loop: refactoring

HashidaTKS commented on Sazabi.cpp:
```diff
@@ -2250,7 +2259,22 @@ BOOL CSazabi::PumpMessage()
+				}
+				else
+				{
+					m_pCefMessageLoopWorker->DoWork();
```
将来的にはこのDoWork()は実行しなくても良いようにしたい。
そうできれば、この上部のキーボードイベント関係のチェックをなくせる。

HashidaTKS commented on SazabiCefMessageLoopWorker.cpp:
```diff
@@ -0,0 +1,85 @@
+	{
+		// Execute the remaining work as soon as possible.
+		PostMessage(m_hWnd_, WM_SCHEDULE_CEF_WORK, NULL, 0);
+	}
```
https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc#L82

参考にした実装だと、ここで保留中でなければIntMax秒のタイマーを起動しようとする。
正直なところ参考の実装でIntMax秒のタイマーを起動している意図は図りかねているのだが、タイマー設定時やタイムアウト時に必ずKillTimerされているので、必要なさそうと判断し、本実装ではその処理はいれていない。

HashidaTKS replied:
https://github.com/ThinBridge/Chronos/issues/162 を考えていて、何のためにサンプルではIntMax秒のタイマーを入れているのか理解できた気がします。

https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc#L82

でIntMax秒のタイマーを設定すると、その後

https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc#L58

で1000/30ミリ秒（30fps）に変更されてタイマーが設定される。
これは何か処理を実行した後、保留中でなければ常に設定されるので、つまり30fpsを担保するために実行されている。
これは、30fpsで`OnScheduleDelayedWork()`が実行されるということでもあり、30fpsで`DoMessageLoopWork()`が実行されるようになっている。
こうすることで、Chronosの場合は、オリジナルのメッセージループで毎回`DoMessageLoopWork()`をする必要がなくなり、`WM_SCHEDULE_CEF_WORK`とタイマータイムアウトのときだけ`DoMessageLoopWork()`をすればよくなる。

ashie replied:
確かにこれは意図がよく分からんですね...

ashie replied:
あ、なるほど、そうですね。

ashie replied:
上記のように修正された

HashidaTKS commented on SazabiCefMessageLoopWorker.cpp:
```diff
@@ -0,0 +1,85 @@
+void SazabiCefMassageLoopWorker::SetTimer(int64_t delayMs)
+{
+	m_bTimerPending_ = true;
+	::SetTimer(m_hWnd_, m_nTimerID, static_cast<UINT>(delayMs), nullptr);
```
* このクラスは常にメッセージループを実行しているメインスレッドから呼ばれる
* メインスレッドのメッセージループの中からしか呼ばれない
* `m_hWnd`はメインスレッドのメインウィンドウを指している

という前提だと、`m_hWnd`は常に正しいウィンドウを指しているはずなので、正しいウィンドウハンドルかどうかのチェックをしていない。

ashie replied:
↑コメントで書くべき内容かなという気もするが、まあコードで分かる話ではあるのでいらないかな。

HashidaTKS commented on Sazabi.cpp:
```diff
@@ -584,6 +586,8 @@ BOOL CSazabi::InitInstance()
 	m_hProcess.Attach(::OpenProcess(PROCESS_ALL_ACCESS, FALSE, m_dwProcessId));
 	InitProcessSetting();
+	m_nCefMessageLoopWorkerTimerId = 1;
+	m_pCefMessageLoopWorker = new SazabiCefMassageLoopWorker(m_pMainWnd->m_hWnd, m_nCefMessageLoopWorkerTimerId);
```
`m_pMainWnd`は`CMainFrm`なので、`CMainFrm`のTimer関係のIDとの重複が無いように注意する必要がある。
（同じウィンドウへのタイマーのタイマーIDは重複してはいけない）

https://github.com/ThinBridge/Chronos/blob/master/MainFrm.h#L104-L106
https://github.com/ThinBridge/Chronos/blob/master/MainFrm.cpp#L120

を見ると、`CMainFrm`のインスタンスのメモリ上のアドレス、アドレス+40、アドレス+48をタイマーIDとして使用しているので、基本的に重複することはなさそうである（アドレスが1でなければ重複しない）。
何故`CMainFrm`側のタイマーIDをアドレスを元に作成しているのかは分からない。もし理由があるなら知りたい。

より厳密にやるなら以下のいずれかの対応の方が良いか。

* `m_nCefMessageLoopWorkerTimerId`に相当するものを`CMainFrm`側に定義して、アドレスを元に作成、そしてそれを持ってくる
* `m_nCefMessageLoopWorkerTimerId`は`1`のままで、`CMainFrm`側のタイマーIDを固定値にしてしまう。ウィンドウ単位でユニークなら良いので、固定値で問題ないはず。
  * その場合でも、`CMainFrm`側に定義があった方が良いか。

ashie replied:
> （同じウィンドウへのタイマーのタイマーIDは重複してはいけない）

ということであれば、タイマーIDはそのウインドウを管理するクラスで管理すべきのように見えます。
つまり、`m_nCefMessageLoopWorkerTimerId`は`m_pMainWnd`が所有すべきもののように見えます。
重複していないかを確認するためにあちこちのソースを確認するのではなく、そのウインドウを管理するクラスだけ見れば良いようにした方が自然な気がします。

ashie replied:
上記のように修正された

HashidaTKS commented:
@ashie 

お手数ですが、再レビューをお願いいたします。

HashidaTKS commented:
https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc#L1

[SazabiCefMessageLoopWorker.cpp](https://github.com/ThinBridge/Chronos/pull/161/files#diff-87fc5721b001d56bb8dea3c73dccb6294df0e99d054385bed327f6c5656cf689)に上記の参考元のライセンス表記を記載すべきかどうかを悩んでいる。
頒布対象であるChronosのバイナリには、バージョン情報のダイアログにCEFのライセンスを記載しているから不要？

HashidaTKS commented on SazabiCefMessageLoopWorker.cpp:
```diff
@@ -0,0 +1,85 @@
+	m_bTimerPending_ = false;
+	m_bIsActive_ = false;
+	m_bReentrancyDetected_ = false;
+	m_nTimerID = nTimerID;
```
一応同じウィンドウに複数のタイマーを動作させられるように、タイマーIDも外部から与えられるようにしている。

HashidaTKS commented on SazabiCefMessageLoopWorker.cpp:
```diff
@@ -0,0 +1,85 @@
+#include "stdafx.h"
+#include "SazabiCefMessageLoopWorker.h"
+
+SazabiCefMassageLoopWorker::SazabiCefMassageLoopWorker(HWND hTargetWnd, UINT_PTR nTimerID)
```
このクラスは一つのスレッドから呼ばれることを想定しているので、このクラス内での排他制御はしていない。

HashidaTKS commented:
> というわけでして、余計な`CefDoMessageLoopWork`が実行されないようにするのは別のPRで検証して、このPRではタイマーの実装までとしたいのですがいかがでしょうか？ （タイマー周りは現在実装中です。）

Issue起票しました。
https://github.com/ThinBridge/Chronos/issues/162

ashie commented on Sazabi.vcxproj:
```diff
@@ -193,6 +193,7 @@
     <ClCompile Include="MainFrm.cpp" />
     <ClCompile Include="MyComboBoxEx.cpp" />
     <ClCompile Include="Sazabi.cpp" />
+    <ClCompile Include="SazabiCefMessageLoopWorker.cpp" />
```
新たに書き起こすソースで`Sazabi`という固有名詞は使いたくないです。
（現在の名前は`Chronos`であり、`Sazabi`って何？となるので、将来的には全て廃止したい）
今回の場合、`Sazabi`は単に削除で良さそうですし、`Cef`も省略して単に`MessageLoopWorker`で十分かなぁという気分です。

ashie replied:
修正された

ashie commented on SazabiCefMessageLoopWorker.cpp:
```diff
@@ -0,0 +1,85 @@
+	}
+}
+
+bool SazabiCefMassageLoopWorker::PerformMessageLoopWork()
```
`MessageLoopWorker`なので関数名のsuffix`MessageLoopWork`は冗長かなぁと思います。
となると単に`Perform`になりますが、`DoWork`と何が違うの？となりそうです。
答えとしては、この場合一つに統合しちゃって良さそうに見えます。

HashidaTKS committed:
rename SazabiCefMassageLoopWorker to MassageLoopWorker

HashidaTKS committed:
Integrate PerformMessageLoopWork() to DoWork()

HashidaTKS committed:
message loop: keep sending the `WM_SCHEDULE_CEF_WORK` message at 30fps

HashidaTKS committed:
message loop: move MessageLoopTimerID into CMainFame

ashie commented:
> https://github.com/chromiumembedded/cef/blob/5e616b2df0811f05a4acad9d2808ad872d2fbd13/tests/shared/browser/main_message_loop_external_pump.cc#L1
> 
> [SazabiCefMessageLoopWorker.cpp](https://github.com/ThinBridge/Chronos/pull/161/files#diff-87fc5721b001d56bb8dea3c73dccb6294df0e99d054385bed327f6c5656cf689)に上記の参考元のライセンス表記を記載すべきかどうかを悩んでいる。 頒布対象であるChronosのバイナリには、バージョン情報のダイアログにCEFのライセンスを記載しているから不要？

これにコメントするの忘れてました。

今回のケースについて言うと、著作権を侵害するレベルで被るコードがあるようには見えませんでした。
ただ私ならば、明らかにこれを参考にしたなぁというコードがある場合には、「ここを参考にしました」という参照URLを記載して、そのライセンスはこうですというのをコメントして記載するかなぁとおもいました。

HashidaTKS committed:
Rename m_pCefMessageLoopWorker to m_pMessageLoopWorker

ashie commented on Sazabi.cpp:
```diff
@@ -2212,9 +2219,9 @@ BOOL CSazabi::PumpMessage()
 				//判明したため、フィルタリングする。multi_threaded_message_loopを有効にすれば問題ないことも判明したが
 				//キーボード、マウス系のイベントがBroViewやBroFrameに飛んでこない弊害あり
-				if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE))
+				if (::PeekMessage(&msg, NULL, NULL, NULL, PM_NOREMOVE))
```
`PeekMessage`の使い方で2点モヤっとしてます。

* 第3,4引数で`NULL`を指定してすべてのメッセージを取っていますが、無駄が多そう
  * ついでに、やるなら引数の型は`UINT`なので`NULL`ではなく`0`かな
  * `PM_NOREMOVE`なので別途これ用に`PeekMessage`実行するか、あるいは（後述の理由もあって）メッセージ受け取るウインドウのウインドウ関数の方で実行した方が良いかも。
* `PeekMessage`って呼ぶの一回で良いんだっけ？
  * 普通は`while (PeekMessage(...))`でキューが捌けるまで回す認識
  * そもそも`CWinThread::PumpMessage`がどういう実装になってるのかは知らないので、正しいオーバーライドの仕方もあまり情報が見つからない（[公式ドキュメントでは上級者のみ推奨と書いてあるのみ](https://learn.microsoft.com/en-us/cpp/mfc/reference/cwinthread-class?view=msvc-170#pumpmessage)）ので何が正しいのか良くわかんない
  *  取りこぼしがあっても良い場合は一回で良いが、取りこぼすと困るやつは真面目に回した方が良いかも。そもそも`PumpMessage`をオーバーライドしてまで無理やりやる処理は少ないほうが良いので、受け取るウインドウが特定されている場合は`PumpMessage`でやるよりそのウインドウのウインドウ関数でやった方が良いかも

ashie replied:
> 第3,4引数でNULLを指定してすべてのメッセージを取っていますが、無駄が多そう

いや、違うか。一回しか実行してないので、無駄が多いと言うより、取りこぼすケースがあるかもしれない、か。

> * `PeekMessage`って呼ぶの一回で良いんだっけ？ 

30fpsの件も結局入れたのであれば、`PeekMessage()`を`WM_TIMER`と`WM_SCHEDULE_CEF_WORK`に対して別途個別に実行するのでひとまずは事足りるか？（`if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE))`の部分には手を付けない。）

ashie replied:
> というわけでして、余計なCefDoMessageLoopWorkが実行されないようにするのは別のPRで検証して、このPRではタイマーの実装までとしたいのですがいかがでしょうか？

これって、上記の理由でメッセージ取りこぼしてるからだったりしませんかね？

HashidaTKS replied:
おっしゃる通りで、ここの使い方はおかしかったため見直し中です。
私の無理解から、一度だけ`PeekMessage`すればいいのかと思っていました！

HashidaTKS replied:
`CMainFrame`に`m_iMessageLoopTimerID`を移したあたりから、そもそも`MessageLoopWorker`は`CMainFrame`にいるべきでは？とも思い始めていたため、そのあたりも含めて考え直しています。

HashidaTKS replied:
> > というわけでして、余計なCefDoMessageLoopWorkが実行されないようにするのは別のPRで検証して、このPRではタイマーの実装までとしたいのですがいかがでしょうか？
> 
> これって、上記の理由でメッセージ取りこぼしてるからだったりしませんかね？

恐らく、おっしゃる通り、メッセージの取りこぼしが理由だと思います。

ashie replied:
> `CMainFrame`に`m_iMessageLoopTimerID`を移したあたりから、そもそも`MessageLoopWorker`は`CMainFrame`にいるべきでは？とも思い始めていたため、そのあたりも含めて考え直しています。

私もちょっと検討してますが、`CefDoMessageLoopWork()`はメッセージ毎に実行するような性質のものではなく、ある程度の時分割毎に実行するもの、と考えると、結局`PumpMessage`にいたほうが良いという考え方もあるかもしれません。そこは実装してみてシンプルあるいは自然になるほうに寄せて良いかと思います。

あと、結局メッセージキューが複数積まれている状況を考慮すると、`if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE)`を削除するのは無理そうな気がしています。
（アップストリームで対処されているとかで問題が発生しなくなっている可能性も無きにしもあらずだけど、コードを引き継いだだけの我々は実際に事象を確認しているわけではないので詳細がよくわからず、下手に手を付けられない）

HashidaTKS replied:
`CMainFrame`側に`MessageLoopWorker`を移動しました。

* `CMainFrame::PreTranslateMessage()`で`WM_SCHEDULE_CEF_WORK`であれば`MessageLoopWorker::OnScheduleWork()`を実行
* `CMainFrame::OnTimer()`でイベントIDが`m_iMessageLoopTimerID`であれば`				m_pMessageLoopWorker->OnTimerTimeout();`を実行
* `MessageLoopWorker`にて、30fpsになるように1/30秒に一度`WM_SCHEDULE_CEF_WORK`を繰り返し送信

となっています。
この動きでは少なくとも`CSazabi::PumpMessage`の中で`if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE)`を見ても意味がないので消してしまっているのですが（`CMainFrame::PreTranslateMessage()`は`CSazabi::PumpMessage()`の後で発生するイベントだから）、確かにメッセージキューが積まれている状態を考えると、どこかで`if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE)`に相当する確認をしなければいけないようにも思えるので、もう少し動作確認しつつ探ってみます。

HashidaTKS replied:
> 少なくとも`CSazabi::PumpMessage`の中で`if (::PeekMessage(&msg, NULL, WM_KEYFIRST, WM_KEYLAST, PM_NOREMOVE)`を見ても意味がないので消してしまっているのですが（`CMainFrame::PreTranslateMessage()`は`CSazabi::PumpMessage()`の後で発生するイベントだから）

考え直してみると、`CSazabi::PumpMessage()`で`CMainFrame`の何かの変数を変更する、といった方法で`CefDoMessageLoopWork`を一旦スキップするように知らせることもできるので、必ずしも`CMainFrame::PreTranslateMessage()`が`CSazabi::PumpMessage()`の後だからといってこの処理を削除する必要はなかったですね。。。

ashie replied:
どうだろ、これで（この処理消して）良い気がしてきた

HashidaTKS replied:
masterブランチで`CSazabi::PumpMessage()`にある「Windows 10 (2004)以降に搭載されているMS-IMEでアドレスバーでEnterキーが効かない問題が発生」について検証してみました。masterブランチで`CSazabi::PumpMessage()`を以下のようにメッセージフィルターがないように改変して動作検証しました。

```
BOOL CSazabi::PumpMessage()
{
	MSG msg = {0};
	TCHAR classname[32] = {0};
	__try
	{
		if (m_bToBeShutdown)
		{
			UnInitializeCef();
			return FALSE;
		}

		//SZB
		if (m_bCEFInitialized)
		{
			if (!m_bMultiThreadedMessageLoop)
			{
				CefDoMessageLoopWork();
			}
		}
		return CWinApp::PumpMessage();
	}
	__except (EXCEPTION_EXECUTE_HANDLER)
	{
		return FALSE;
	}
}
```

この状態で元のソースコードのコメントにあるようにアドレスバーに文字列を入れてエンターを押すと、正しく動作しました（再現しない）。
その後、

```
	settings.external_message_pump = true;
```

を

```
	settings.external_message_pump = false;
```

に変更し、再度アドレスバーに文字列を入れてエンターを押すと、正しく動作しなくなりました（再現した）。

つまり、実は「Windows 10 (2004)以降に搭載されているMS-IMEでアドレスバーでEnterキーが効かない問題が発生」は`settings.external_message_pump = true;`としたタイミングでフィルタリングなしでも正しく動作するように直っていたのではないかと思います。

そのため、実は既にこの箇所は不要なので、消してしまって大丈夫なのではないかと思います。
（別途この処理を削除するところだけ、別のPRに分けた方が良いかもしれませんが）

ashie replied:
なるほど、まあメッセージループの改善という括りで一緒で良いと思います！

HashidaTKS committed:
message loop: fix the position of executing MessageLoopWorker

HashidaTKS committed:
message loop: add a comment about a reference source code

HashidaTKS committed:
fix typos

HashidaTKS committed:
message loop: move MessageLoopWorker into CMainFrame

HashidaTKS committed:
message loop: remove a needless message check

HashidaTKS committed:
CMainFrame: add process to cleanup m_pMessageLoopWorker

HashidaTKS committed:
MainFrm.cpp: add a newline

ashie commented on MainFrm.cpp:
```diff
@@ -2822,3 +2851,21 @@ void CMainFrame::OnSettingChange(UINT uFlags, LPCTSTR lpszSection)
+				if (!theApp.m_bMultiThreadedMessageLoop)
+				{
+					int64_t delayMs = pMsg->lParam;
+					m_pMessageLoopWorker->OnScheduleWork(delayMs);
```
これ、このメッセージについては処理済みなので、`TRUE`返してこれ以上処理させないようにした方が良いのでは？
https://learn.microsoft.com/ja-jp/cpp/mfc/reference/cwnd-class?view=msvc-170#return-value-170
誰も処理しないと思うので実質無害だとは思いますが、`CFrameWnd::PreTranslateMessage(pMsg)`呼ぶ必要は無いと思います。

HashidaTKS replied:
おっしゃるとおりです。修正します。

ashie replied:
あとネストが深い...`if`分一個で条件を`&&`で繋げるだけで良さそうです。

HashidaTKS replied:
ネストを減らしつつ`WM_SCHEDULE_CEF_WORK`を処理したら`TRUE`を返すように修正しました。
後続でこのメッセージを使うことはないため、`m_pMessageLoopWorker->OnScheduleWork`を実行しなかった場合でも`TRUE`を返しています。

HashidaTKS committed:
message loop: return immediately after processing WM_SCHEDULE_CEF_WORK.

HashidaTKS committed:
Reduce nests

HashidaTKS commented:
一旦、ご指摘頂いた事項には全て対応したつもりでおります。

ashie commented:
> 一旦、ご指摘頂いた事項には全て対応したつもりでおります。

あざっす、だいぶ良くなった気がします。
このままでも良いと思うのですが、週明けマージ前にもう一息だけで、以下の点を軽く検討してみてもらえないでしょうか。

`CSazabi::PumpMessage()`は完全に消せないか？
というかここでの`UnInitializeCef()`をやめてもっとまっとうな箇所でやれないか？
というか、`ExistInstance()`でもしつこくやってるが、こんなに何回もやる必要があるのか？

検討に時間がかかりそうな匂いがしたら今回は軽く確認する程度にしておいて、後々ちゃんと検討するでも良いですし、明らかに消せそうなら今回消しちゃっても良いです。

ashie commented on Sazabi.cpp:
```diff
@@ -2201,58 +2201,6 @@ BOOL CSazabi::PumpMessage()
 			UnInitializeCef();
```
ローカル変数の`msg`と`classname`が不要になっています。

ashie commented:
> `CSazabi::PumpMessage()`は完全に消せないか？ というかここでの`UnInitializeCef()`をやめてもっとまっとうな箇所でやれないか？ というか、`ExistInstance()`でもしつこくやってるが、こんなに何回もやる必要があるのか？

私の環境で軽く確認してみましたが、単純に`CSazabi::PumpMessage()`を消した方が調子が良さそうです。
これがあると、ATOKを使用しているときにChronosの終了コードが`1`になるという問題がありましたが、消すと問題無く`0`で終了するようになりました（MS-IME、ATOK、Google日本語入力で試しましたが、ATOKのみこの問題が発生します）。

あと、 #163 がこの修正と組み合わせたときに本当に必要になるか、今一度確認してみてもらえないでしょうか？
私のところでは、軽く確認する限り #163 が無くても問題無いように見えます。

ashie commented:
さらにテストしていたところ、大きな問題を見つけました。
ウインドウをリサイズすると、画面が更新されなくなります。

HashidaTKS commented:
取り急ぎ、`CSazabi::PumpMessage`は削除しました。

HashidaTKS commented:
確かに、リサイズすると画面が反応しなくなりました。
`MessageLoopWorker`が動かなくなっているので、メッセージが飛んでこなくなっている気配があります。

ashie commented:
> 確かに、リサイズすると画面が反応しなくなりました。 `MessageLoopWorker`が動かなくなっているので、メッセージが飛んでこなくなっている気配があります。

問題のコミットを特定しました。
~~b5dfb5a0c40437e0b137531692f00a2080459f9d 以降でリサイズが動かなくなります。~~
いや、違うな...
35dd703b36218146d119b676b790d08f6405a429 は動かない
b5dfb5a0c40437e0b137531692f00a2080459f9d は動く
たぶん 07adac069f7985c363fb6ca66aca3d894b52fb49 ですね

HashidaTKS commented:
ありがとうございます！

HashidaTKS commented:
リサイズのタイミング以降、`MessageLoopWorker::DoWork()`で`PostMessage(m_hWnd_, WM_SCHEDULE_CEF_WORK, NULL, static_cast<LPARAM>(m_nTimerDelayPlaceholder));`を送ってから、次に`OnScheduleWork()`が実行されていない。

ログを仕込んで様子を見ると、ウィンドウハンドルは変わっていないのだが、何故メッセージが到達しなくなってしまったのだろう。。。

HashidaTKS commented:
アクティブでないウィンドウにメッセージは到達しない？
そのためリサイズ時に一時的にメッセージが無視されて、結果的にループがそこで止まってしまっている？

HashidaTKS commented:
メッセージ処理専用のウィンドウを作れば解決するのだろうか？

ashie commented:
詳細はよくわかってませんが、たしかに`CefDoMessageLoopWork()`の実行チェーンが途切れて止まっているように見えます。
`CSazabi::PumpMessage()`なり`CMainFrame::PreTranslateMessage()`なりで無理やり`CefDoMessageLoopWork()`を実行するようにしていると、止まらずに復帰する。

HashidaTKS commented:
メッセージ処理専用のウィンドウでMessageLoopWorkerを動かせば問題ないことが判明したため、その方針で対応中です。

HashidaTKS committed:
message loop: move message loop into a message exclusive window

ashie commented on MainFrm.cpp:
```diff
@@ -167,7 +167,6 @@ void CMainFrame::View_InitOK()
 			m_iRecoveryTimerID = (INT_PTR)this + 40;
 			this->SetTimer(m_iRecoveryTimerID, 15 * 1000, 0);
 		}
-
```
この辺の余分な差分が出ないように修正してもらえますか？

ashie replied:
修正された

HashidaTKS replied:
すみません、その他の余分な差分も含めて見直し途中でした。。。

HashidaTKS committed:
Revert MainFrm.cpp/h changes

HashidaTKS committed:
Remove a needless newline

HashidaTKS committed:
Remove a needless newline

ashie commented on Sazabi.cpp:
```diff
@@ -4318,6 +4266,65 @@ void CSazabi::UnInitializeCef()
 	}
 }
+
+void CSazabi::InitMessageLoopWindow()
```
このウインドウ、`MessageLoopWorker`に持たせたほうがほうがすっきりしないかな？

HashidaTKS replied:
ウィンドウはCSazabiが管理するものというイメ―ジが何故か強くあってCSazabiに持たせていたのですが、確かにMessageLoopWorkerに持たせた方がすっきりしそうです。見直します。

ashie replied:
直った

ashie commented on client_app.cpp:
```diff
@@ -131,6 +131,16 @@ void ClientApp::OnBeforeCommandLineProcessing(const CefString& process_type, Cef
+	if (!theApp.SafeWnd(theApp.m_pMainWnd->m_hWnd))
+		return;
+
+	theApp.m_pMainWnd->PostMessage(WM_SCHEDULE_CEF_WORK, NULL, static_cast<LPARAM>(delayMs));
```
`PostMessage`の送信先を変える必要がありますね。
`MessageLoopWorker`にそれ用のメンバー関数持たせると良いかも。

ashie replied:
直った

HashidaTKS committed:
message loop: move the message loop window into MessageLoopWorker

HashidaTKS committed:
Remove needless statements

HashidaTKS committed:
message loop: modify OnScheduleMessagePumpWork PostMessage target

HashidaTKS committed:
commit uncommitted changes

HashidaTKS committed:
message loop: rename MessageLoopWindowHandler to WindowProcesser

HashidaTKS committed:
message loop: fix hWnd check in OnScheduleMessagePumpWork

HashidaTKS committed:
Fix format

HashidaTKS commented on MessageLoopWorker.cpp:
```diff
@@ -0,0 +1,165 @@
+						    WPARAM wparam,
+						    LPARAM lparam)
+{
+	if (!theApp)
```
あまりここで`theApp`を参照したくはない。
`m_bCEFInitialized`を本当に見る必要があるのか、というところも疑問。普通のシナリオなら、`MessageLoopWorker`の動作中は常に`m_bCEFInitialized`であるはず。仮にこのフラグ値を見る必要があるとして、MessageLoopWorkerの中にフラグをもって、`m_bCEFInitialized = false`という代入が行われる場所で、そのフラグを変更する方が良くないか？という気もする。

ashie replied:
その点も含めて（たぶん最後の）コード整理のご提案。
だいぶコードがあるべき姿になってきて、今回の問題の責務を`MessageLoopWorker`に集約すれば良さそうということが見えてきていると思います。
そこで以下のようにするとどうでしょうか？

* たぶん`MessageLoopWorker::MessageLoopWorker()`に引数を追加して`HINSTANCE`だけ渡せば良いと思います。
* `MessageLoopWorker`の初期化は`CSazabi::InitializeCEF()`で、破棄は`CSazabi::UnInitializeCEF()`でやるようにするとどうでしょう。
  * #163 のような処理も`MessageLoopWorker`に持たせて`CefDoMessageLoopWork()`実行の責務を完全に`MessageLoopWorker`に持たせた方が良さそう。
* `MessageLoopWorker::m_hWnd`を公開するよりはメッセージを投げる関数を追加する方が私は好み
   `messageLoopWorker->Schedule(delayMs)`みたいな感じ

必然的にオブジェクトの寿命管理も整理されて、余計なチェックが要らなくなる気がします。

HashidaTKS replied:
>     * たぶん`MessageLoopWorker::MessageLoopWorker()`に引数を追加して`HINSTANCE`だけ渡せば良いと思います。

その通りでした。
`MessageLoopWorker::MessageLoopWorker()`に引数を追加して`HINSTANCE`だけ渡すように修正しました。

>     * `MessageLoopWorker`の初期化は`CSazabi::InitializeCEF()`で、破棄は`CSazabi::UnInitializeCEF()`でやるようにするとどうでしょう。

その方が良さそうでしたので、そのように変更しました。

>       * [Avoid exceptions at cef_log during shutting down #163](https://github.com/ThinBridge/Chronos/pull/163) のような処理も`MessageLoopWorker`に持たせて`CefDoMessageLoopWork()`実行の責務を完全に`MessageLoopWorker`に持たせた方が良さそう。

#163 に相当する処理も`MessageLoopWorker`に入れました。
また、以前言及されていた通り、`CefDoMessageLoopWork`を10回ループせず、一度実行するだけで大丈夫そうでしたので、そのように修正しました。

>     * `MessageLoopWorker::m_hWnd`を公開するよりはメッセージを投げる関数を追加する方が私は好み
>       `messageLoopWorker->Schedule(delayMs)`みたいな感じ

`messageLoopWorker->Schedule(delayMs)`を実行すると、最終的に`::SetTimer(m_hWnd_, m_nTimerID, static_cast<UINT>(delayMs), nullptr);`が実行されるのですが、`SetTimer`で指定するハンドルは、呼び出し元のスレッドが所有している必要があり、複数のスレッドから呼ばれる可能性のある`ClientApp::OnScheduleMessagePumpWork`からは直接呼べなさそうです。

https://learn.microsoft.com/ja-jp/windows/win32/api/winuser/nf-winuser-settimer#parameters

> タイマーに関連付けるウィンドウへのハンドル。 このウィンドウは、呼び出し元のスレッドによって所有されている必要があります。

`PostMessage`にはそのような制約はないため、`MessageLoopWorker`向けに`PostMessage`を抽象化した`MessageLoopWorker::PostScheduleMessage(int64_t delayMs)`を作成して、そちらを使うようにしました。
複数のスレッドから呼ばれる可能性のある`ClientApp::OnScheduleMessagePumpWork`から`MessageLoopWorker::PostScheduleMessage(int64_t delayMs)`を呼んでいるので、厳密にはその内部で呼び出されている`m_hWnd_`で競合が発生しないように排他を考えなければいけないと思うのですが、`m_hWnd_`が変更されるのはChronosの初期化時と終了時だけなので、大丈夫ではないかと思い一旦排他しないでいます。

ashie replied:
ああ、すみません。意図としてはその関数の中で`PostMessage`するというつもりで書いてました。
なので仰る通りです。

ashie replied:
> 大丈夫ではないかと思い一旦排他しないでいます。

はい、大丈夫だと思います

HashidaTKS committed:
message loop: add PostScheduleMessage function

HashidaTKS committed:
Rename m_hWnd to m_hWnd_ because it changed to be private

HashidaTKS committed:
message loop: move MessageLoopWorker initializing/unInitializing into CEF scope

HashidaTKS committed:
Use bool instead of BOOL

HashidaTKS commented on Sazabi.cpp:
```diff
@@ -4305,14 +4241,11 @@ void CSazabi::UnInitializeCef()
 		m_bCEFInitialized = FALSE;
 		m_cefApp = nullptr;
-		if (!m_bMultiThreadedMessageLoop)
+		if (m_pMessageLoopWorker)
```
`!m_bMultiThreadedMessageLoop`の時しか`m_pMessageLoopWorker`は初期化されないので、`m_pMessageLoopWorker`の存在を確認するだけで充分。

HashidaTKS commented on Sazabi.cpp:
```diff
@@ -4305,14 +4241,11 @@ void CSazabi::UnInitializeCef()
-				// Sleep to allow the CEF proc to do work.
-				Sleep(50);
-			}
+			m_pMessageLoopWorker->Quit();
```
`delete`すればデストラクタが呼ばれて、その中で`Quit()`が呼ばれるので、`delete`するだけで十分なのだが、一応`Run()`に対応して明示的に`Quit()`を定義し、それを呼びだす方がそれっぽい気がしたのでこのようにした。

HashidaTKS commented on client_app.cpp:
```diff
@@ -131,6 +131,14 @@ void ClientApp::OnBeforeCommandLineProcessing(const CefString& process_type, Cef
+	MessageLoopWorker* messageLoopWorker = theApp.m_pMessageLoopWorker;
+	if (!messageLoopWorker)
+		return;
+	messageLoopWorker->PostScheduleMessage(delayMs);
```
`OnScheduleMessagePumpWork`がマルチスレッドで呼ばれる可能性があるので、`PostScheduleMessage`もマルチスレッドで呼ばれる可能性がある点に注意。
なので、厳密には`MessageLoopWorker::m_hWnd`の参照と変更を排他制御しないといけない気がする。
とはいえ、`MessageLoopWorker::m_hWnd`が変更されるのはChronosの初期化時と終了時のみで、現実的に問題になることはなさそうなのだが。

ashie reviewed (approved):
本日変更を逐次追いながら一通り動作を確認していましたが、 d493cb895730daef7aa25a5ac99ce766b4cdf1a0 で全ての問題が改善していると思いますし、コード上の不可解だった点・気持ち悪かった点も全て解消していると思います。
今回改めて動作確認していたところ、修正前は色々挙動不審なところがあってプロセスが生き残る問題がまだ再現することもありましたが、修正後は解消していると思います。
心なしか、動作も以前よりキビキビしているように感じます。

明日、もう少し軽く確認したらマージしようと思います。
長々とご苦労様でした！

ashie commented:
すみません、やっぱり @HashidaTKS さんの方でマージしてください。私の方は現状でOKです。
Squash mergeするのに、コミットメッセージを自分の言葉でまとめてもらった方が良いかと思いますので。

HashidaTKS commented:
レビューありがとうございました！マージしました！
